cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -1400,21 +1400,70 @@ VersionedItem LocalVersionedEngine::append_internal(
     }
 }
 
-std::vector<AtomKey> LocalVersionedEngine::batch_append_internal(
-    std::vector<VersionId> version_ids,
+std::vector<std::variant<VersionedItem, DataError>> LocalVersionedEngine::batch_append_internal(
     const std::vector<StreamId>& stream_ids,
-    std::vector<AtomKey> prevs,
-    std::vector<InputTensorFrame> frames,
-    const WriteOptions& write_options,
-    bool validate_index) {
+    std::vector<InputTensorFrame>&& frames,
+    bool upsert,
+    bool prune_previous_versions,
+    bool validate_index,
+    bool throw_on_missing_version) {
 
-    std::vector<folly::Future<AtomKey>> append_futures;
-    for(auto id : folly::enumerate(stream_ids)) {
-        UpdateInfo update_info{prevs[id.index], version_ids[id.index]};
-        append_futures.emplace_back(async_append_impl(store(), update_info, std::move(frames[id.index]), write_options, validate_index));
+    auto write_options = get_write_options();
+    auto stream_update_info_futures = batch_get_latest_undeleted_version_and_next_version_id_async(store(),
+                                                                                                    version_map(),
+                                                                                                    stream_ids);
+    std::vector<folly::Future<VersionedItem>> append_versions_futs;
+    internal::check<ErrorCode::E_ASSERTION_FAILURE>(stream_ids.size() == stream_update_info_futures.size(), "stream_ids and stream_update_info_futures must be of the same size");
+    for (const auto&& [idx, stream_update_info_fut] : folly::enumerate(stream_update_info_futures)) {
+        append_versions_futs.push_back(
+            std::move(stream_update_info_fut)
+            .thenValue([this, frame = std::move(frames[idx]), write_options = std::move(write_options), validate_index, upsert](auto&& update_info) mutable -> folly::Future<IndexKeyAndUpdateInfo> {
```

Comment:
Note that async_append_impl also uses write_options, but I believe it can still be refactored in a way that allows us to obtain write_options only once.